From: Mirakels <mirakels@openpli.org>
Date: Thu, 20 Mar 2014 21:46:03 +0100
Subject: [PATCH] Handle DTS Passthrough and Downmix

reworked pathc from Athanasios Oikonomou <athoik@gmail.com>
to cope with latest upstream code:

Check for ac3plus using access() system call.
If downmix is enabled remove DTS caps to give priority on DTS downmix.

diff --git a/src/gstdvbaudiosink.c b/src/gstdvbaudiosink.c
index 80234c0..ebd9ef2 100644
--- a/src/gstdvbaudiosink.c
+++ b/src/gstdvbaudiosink.c
@@ -303,25 +303,25 @@ gst_dvbaudiosink_get_caps (GstBaseSink *basesink)
 //	GstDVBAudioSink *self = GST_DVBAUDIOSINK (basesink);
 //	gchar *strcaps;
 	GstCaps *caps;
-	static int eac3_support;
 
-	if (!eac3_support) {
-		int fd = open("/proc/stb/audio/ac3plus", O_RDONLY);
+	caps = gst_caps_copy(&hwtemplate->static_caps.caps);
+	if (access("/proc/stb/audio/ac3plus", R_OK)) {
+		gst_caps_remove_structure(caps, 12); // remove x-eac3
+		gst_caps_remove_structure(caps, 12); // remove x-private-eac3
+	}
+
+	if (hwtemplate == &sink_factory_broadcom_dts) {
+		int fd = open("/proc/stb/audio/ac3", O_RDONLY);
+		char buf[32] = "";
 		if (fd >= 0) {
-			eac3_support = 1;
+			read(fd, buf, 32);
 			close(fd);
 		}
-		else
-			eac3_support = -1;
-	}
-
-	if (eac3_support < 0) {
-		caps = gst_caps_copy(&hwtemplate->static_caps.caps);
-		gst_caps_remove_structure(caps, 12); // remove x-eac3
-		gst_caps_remove_structure(caps, 12); // remove x-private-eac3
+		if (!strncmp(buf, "downmix", 7)) {
+			gst_caps_remove_structure(caps, 9); // remove x-private1-dts
+			gst_caps_remove_structure(caps, 8); // remove x-dts
+		}
 	}
-	else
-		caps = gst_static_caps_get(&hwtemplate->static_caps);
 
 //	strcaps = gst_caps_to_string(caps);
 //	GST_INFO_OBJECT (self, "dynamic caps for model %d '%s'", hwtype, gst_caps_to_string(caps));
